File: src/PlusOne.java
import java.util.ArrayList;

/**
 * 
 * PlusOne.java
 * Dec 1, 2012
 */

/**
 * @author huangsp
 *
 */
public class PlusOne {
	public int[] plueOne(int[] num){
		ArrayList<Integer> sum = new ArrayList<Integer>();
		if (num.length == 0) 	return null;
		else{
			int index = num.length - 1;
			int carry = 0;
			for (; index >= 0; index--){
				int sum1 = num[index] + carry;
				carry = sum1 / 10;
				int digit = sum1 % 10;
				sum.add(digit);
			}
			int[] solution = new int[sum.size()];
			for (index = sum.size() - 1; index >=0; index--){
				solution[sum.size() - 1 - index] = sum.get(index);
			}
			return solution;
		}
	}
}

File: python/plus_one.py
#! /usr/bin/python

'''
Given a non-negative number represented as an array of digits, plus one to the number.

The digits are stored such that the most significant digit is at the head of the list.
'''

class Solution:
    # @param digits, a list of integer digits
    # @return a list of integer digits
    def plusOne(self, digits):
        size = len(digits)
        carry_on = 1
        for index in range(size - 1, -1, -1):
            sum = digits[index] + carry_on
            carry_on = sum / 10
            digits[index] = sum % 10
            if carry_on == 0:
                break
        if carry_on == 1:
            digits.insert(0,1)
        return digits

if __name__ == '__main__':
    solution = Solution()
    print solution.plusOne([9,9,9])
    print solution.plusOne([0])
    print solution.plusOne([])
    print solution.plusOne([9,8,9])

